js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+          return length;
+        }
+        if (flexbuffers.ValueType.isFixedTypedVector(valueType)) {
+          length = flexbuffers.ValueType.fixedTypedVectorElementSize(valueType);
+        } else if (valueType === flexbuffers.ValueType.BLOB
+          || valueType === flexbuffers.ValueType.MAP
+          || flexbuffers.ValueType.isAVector(valueType)) {
+          length = readInt(dataView, indirect(dataView, offset, parentWidth) - byteWidth, flexbuffers.BitWidth.fromByteWidth(byteWidth))
+        } else if (valueType === flexbuffers.ValueType.NULL) {
+          length = 0;
+        } else if (valueType === flexbuffers.ValueType.STRING) {
+          const _indirect = indirect(dataView, offset, parentWidth);
+          let sizeByteWidth = byteWidth;
+          size = readInt(dataView, _indirect - sizeByteWidth, flexbuffers.BitWidth.fromByteWidth(byteWidth));
+          while (dataView.getInt8(_indirect + size) !== 0) {
+            sizeByteWidth <<= 1;
+            size = readInt(dataView, _indirect - sizeByteWidth, flexbuffers.BitWidth.fromByteWidth(byteWidth));
+          }
+          length = size;
+        } else if (valueType === flexbuffers.ValueType.KEY) {
+          const _indirect = indirect(dataView, offset, parentWidth);
+          size = 1;
+          while (dataView.getInt8(_indirect + size) !== 0) {
+            size++;
+          }
+          length = size;
+        } else {
+          length = 1;
+        }
+        return length;
+      },
+
+      toObject: function() {
+        const length = this.length();
+        if (this.isVector()) {
+          let result = [];
+          for (let i = 0; i < length; i++) {
+            result.push(this.get(i).toObject());
+          }
+          return result;
+        }
+        if (this.isMap()) {
+          let result = {};
+          for (let i = 0; i < length; i++) {
+            let key = keyForIndex(i, dataView, offset, parentWidth, byteWidth);
+            result[key] = valueForIndexWithKey(i, key, dataView, offset, parentWidth, byteWidth, length, path).toObject();
+          }
+          return result;
+        }
+        if (this.isNull()) {
+          return null;
+        }
+        if (this.isBool()) {
+          return this.boolValue();
+        }
+        return this.numericValue() || this.blobValue() || this.stringValue();
+      }
+    }
+  }
+
+  const len = buffer.byteLength;
+
+  if (len < 3) {
+    throw "Buffer needs to be bigger than 3";
+  }
+
+  const dataView = new DataView(buffer);
+  const byteWidth = dataView.getUint8(len - 1);
+  const packedType = dataView.getUint8(len - 2);
+  const parentWidth = flexbuffers.BitWidth.fromByteWidth(byteWidth);
+  const offset = len - byteWidth - 2;
+
+  return Reference(dataView, offset, parentWidth, packedType, "/")
+};
+
+flexbuffers.toObject = (buffer) => {
+  return flexbuffers.toReference(buffer).toObject();
+};
+
+flexbuffers.builder = (size = 2048) => {
+  let buffer = new ArrayBuffer(size > 0 ? size : 2048);
```

Comment:
Just let the `ArrayBuffer` constructor throw something if someone passes a negative size?

Reply:
Oh there is another purpose in this expression. The grow buffer algorithm will not work properly if the `ArrayBuffer` size is 0. So I would rather keep it here then checking each time I need to grow the buffer. It is also called just once per buffer creation so IMHO no big performance penalty.